### ISSUE TITLE
Consider adding numpy number types to TParamVal

### ISSUE THREAD

**Issue by dabacon** [quantumlib/Cirq#5758]
Current we define
```
TParamVal = Union[float, sympy.Expr]
```
for parameter values (like exponents in `cirq.EigenGate`s).  However in general these also work for numpy types like `np.double`.

We should consider adding support for these numpy types

**Comment by tanujkhattar** [comment]
Also consider moving `Scalar` type from `linear_dict.py` to `type_alias.py`.

**Comment by vtomole** [comment]
If we do the above then `cirq.value.Scalar` won't work anymore. We need to decide whether or not that qualifies as a breaking change.

xref https://github.com/quantumlib/Cirq/issues/5745

**Comment by daxfohl** [comment]
This is somewhat related to #2936. It'd be nice, if we do this, to distinguish between what types are accepted as parameters, and what types are used internally, and ensure that everywhere parameters can be set, we proactively perform those conversions if necessary. Otherwise we have to handle all the different possible numeric types at the point of consumption, which gets unwieldly and leads to backwards compatibility concerns. So, likely differentiating between `SupportsTParamVal`, which is a type that can be converted to a parameter type, and `TParamVal` itself, along with a function or protocol to convert them, would be desirable here.

See https://github.com/quantumlib/Cirq/issues/2936#issuecomment-2770516412 for details.

**Comment by daxfohl** [comment]
Of course, if there's a reason users want to use np types in the implementation of their parameters, that's still perfectly possible: the np type can be added to the `TParamVal` union too. But the distinction between `TParamVal` and `SupportsTParamVal` would allow for a decoupling between what types are supported in the setters vs what types are used internally.

**Comment by ACE07-Sev** [comment]
@daxfohl Greetings,

Hope you're doing well. Anything that needs to be done for this issue?

**Comment by daxfohl** [comment]
@ACE07-Sev I don't know. It hasn't been completed, as you can see in https://github.com/quantumlib/Cirq/blob/main/cirq-core/cirq/value/type_alias.py#L28, `TParamVal` still does not include numpy types. As vtomole mentioned, changing this would be a breaking change. It'd be fine for contravariant (input) use cases like `gate = XPowGate(exponent=np.double(0.5))` but it'd potentially break covariant cases (reads) since `gate.exponent` would then have more possible types to handle. So user code like `is_parameterized = not isinstance(gate.exponent, float)` would become invalid.

The only way to do this without it being a breaking change would be to break the covariant and contravariant cases into separate types. So the covariant cases like `get_exponent` would use the existing `TParamVal`, and the contravariant cases like the constructors would use a new `TParamValInput` that also includes numpy types, and then convert that to `TParamVal` before setting the member variable.

For the case in #2936, the decision was to add code that raises an exception if a string is passed in. That made sense in that case because most use cases would eventually fail anyway if a raw string was used for a parameter. In this case, that wouldn't be a good approach, because there are likely users who do use numpy types even though mypy says they shouldn't, and I expect they'll usually work just fine, so adding an exception would break a bunch of existing user code.

I can't speak for the maintainers (I'm not one), but the only two options I see here are either to add the new `TParamValInput` and a helper `convert(v: TParamValInput) -> TParamVal` and use those in the constructors everywhere, or to close the issue.
